### PR TITLE
EVAKA-4240: citizen makes reservations backend v1

### PR DESCRIPTION
**Pull request by Joosakur**

### CODE REVIEW

**Review by jaahon — COMMENTED** [review]
Good stuff!

File: service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            ReservationRequest(
+                children = setOf(testChild_1.id, testChild_2.id),
+                reservations = listOf(
+                    DailyReservationRequest(monday, startTime, endTime),
+                    DailyReservationRequest(monday.plusDays(1), startTime, endTime)
+                )
+            )
+        )
+
+        val res = getReservations()
```
[the remark this answers is not in the document]
Comment (Joosakur) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenControllerTest.kt:56]: true 👍

File: service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenQueriesTest.kt [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+import fi.espoo.evaka.testChild_1
+import fi.espoo.evaka.testChild_2
+import fi.espoo.evaka.testDaycare
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import java.time.LocalDate
+import java.time.LocalTime
+import kotlin.test.assertEquals
+
+class ReservationCitizenQueriesTest : PureJdbiTest() {
```
[the remark this answers is not in the document]
Comment (Joosakur) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenQueriesTest.kt:23]: Yeah I'm not sure. These probably run a bit faster and the controller's responsibilities get already tested with one test.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt [hunk 3]
```diff
@@ -32,8 +34,101 @@ class ReservationControllerCitizen {
...
+
+    @PostMapping("/citizen/reservations")
+    fun postReservations(
+        db: Database.Connection,
+        user: AuthenticatedUser,
+        @RequestBody body: ReservationRequest
+    ) {
+        Audit.AttendanceReservationCitizenCreate.log(targetId = body.children.joinToString())
+        user.requireOneOfRoles(UserRole.CITIZEN_WEAK, UserRole.END_USER)
+
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt:46]: should we throw Forbidden already here if some child is not user's dependant? probably yes

Comment (jaahon) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt:46]: Maybe, if the response is the same when the child ids don't exist.

File: service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenQueriesTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaahon) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/reservations/ReservationCitizenQueriesTest.kt:23]: My idea was just that the controller api doesn't change as often so refactoring is less painful. But the tests already test a bit higher level function, so these aren't really "query" tests.

**Review by jaahon — APPROVED** [review]

**Pull request merged**